boot/seal.go

```diff
@@ -556,9 +394,66 @@ func modesForSystems(modeenv *Modeenv) map[string][]string {
 	return systemToModes
 }
 
+func recoveryBootChainsForSystems(systems []string, modesForSystems map[string][]string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv, includeTryModel bool, seedDir string) (chains []BootChain, err error) {
+	if trbl == nil {
+		return recoveryBootChainsForSystemsUntrusted(systems, modesForSystems, modeenv, includeTryModel, seedDir)
+	}
+
+	return recoveryBootChainsForSystemsTrusted(systems, modesForSystems, trbl, modeenv, includeTryModel, seedDir)
+}
+
+func recoveryBootChainsForSystemsUntrusted(systems []string, modesForSystems map[string][]string, modeenv *Modeenv, includeTryModel bool, seedDir string) (chains []BootChain, err error) {
+	chainsForModel := func(model secboot.ModelForSealing) error {
+		for _, system := range systems {
+			var cmdlines []string
+			modes, ok := modesForSystems[system]
+			if !ok {
+				return fmt.Errorf("internal error: no modes for system %q", system)
+			}
+
+			for _, mode := range modes {
+				// FIXME: we do not really know the
+				// command line. But we do know the
+				// mode and system we should give that
+				// to the fde manager.
+				switch mode {
+				case ModeRun:
+					cmdlines = append(cmdlines, "snapd_recovery_mode=run")
+				case ModeRecover:
+					cmdlines = append(cmdlines, fmt.Sprintf("snapd_recovery_system=%v snapd_recovery_mode=recover", system))
+				case ModeFactoryReset:
+					cmdlines = append(cmdlines, fmt.Sprintf("snapd_recovery_system=%v snapd_recovery_mode=factory-reset", system))
+				}
+			}
+
+			chains = append(chains, BootChain{
+				BrandID:        model.BrandID(),
+				Model:          model.Model(),
+				Classic:        model.Classic(),
+				Grade:          model.Grade(),
+				ModelSignKeyID: model.SignKeyID(),
+				KernelCmdlines: cmdlines,
+			})
+		}
+		return nil
+	}
+
+	if err := chainsForModel(modeenv.ModelForSealing()); err != nil {
+		return nil, err
+	}
+
+	if modeenv.TryModel != "" && includeTryModel {
+		if err := chainsForModel(modeenv.TryModelForSealing()); err != nil {
+			return nil, err
+		}
+	}
+
+	return chains, nil
+}
+
 // TODO:UC20: this needs to take more than one model to accommodate the remodel
 // scenario
```

Comment:
I think this TODO has been actually implemented by the TryModel support?

Reply:
I think there might be another similar TODO in the file that can also be removed